# Fixed Income Market for Corporate Issuers

### REPO Market calculations

<hr>
    
<b><u>Repurchase Agreements Example</b></u>

Here we don't know the **Initial Margin**
    
A 100 million dollar US 5-year Treasury note sold under a repo, for repurchase in 30 days at a repo rate of 0.25%, with an initial margin required of 102%
    
        * Borrower sells note today, recieving loan of 100,000,000 dollars.
        * In 30 days, borrower will repurchase the note and pay the lender.
    
        * Typically post extra collateral above the loan amount *(Initial Margin)*
            * (i.e., the loan will be at discount to the value of the security)
            
* $\qquad\text{Initial Margin} = \large\frac{\text{security price}}{\text{loan amount}}$

<br>


* $\qquad\text{Repo haircut} = \large\frac{\text{security price - purchase price}}{\text{security price}}$


<hr>

Buyer sells the note today, recieving a loan of:  

In [3]:
# Calculate & Print the Initial Margin required
init_margin = 1.02
sec_price = 100000000
buy_price = sec_price / init_margin


# Calculate Repo haircut
repo_haircut = (sec_price - buy_price) / sec_price


# Calculate 30 day interest on the loan
int_30 = sec_price * (0.0025 / 12)

# Calculate the principal and interest on loan
pr_int = sec_price + int_30


## Print Results
print("Initial Price of Loan  at day 0   = $", round(sec_price, 2))
print("Purchase Price of Loan at day 5   = $", round(buy_price, 0))
print("30 day interest on Loan           = $", round(int_30, 2))
print("Principal + Interest 30 day loan  = $", round(pr_int,2))
print("Repo Haircut of the Loan          = %", round(repo_haircut, 4))


print("")
print("")
print("In 30 days, borrower will repurchase note & pay lender $100,020,833.33")
print("")
print("Loan of $98,039,216 had a 30 day interest of $20,833.33")


Initial Price of Loan  at day 0   = $ 100000000
Purchase Price of Loan at day 5   = $ 98039216.0
30 day interest on Loan           = $ 20833.33
Principal + Interest 30 day loan  = $ 100020833.33
Repo Haircut of the Loan          = % 0.0196


In 30 days, borrower will repurchase note & pay lender $100,020,833.33

Loan of $98,039,216 had a 30 day interest of $20,833.33


Repo Haircut can also be calculated as $\qquad\text{Repo haircut} =\large 1 - (\frac{1}{\text{initial margin}})$

<br>

$\qquad\text{Variation Margin} = \text{(initial margin x adjusted purchase price) - market value of collateral}$

### Example 2: Repo mkt loan with collateral adjustment in 90 days

To protect the lender against a potential decrease in the value of the securities posted as collateral, the borrower typically must post extra collateral above the loan amount (the purchase price) by an amount known as the initial margin. In practice this means the loan amount will be a discount to the value of the securities.


#### Example 2:

Consider a firm that wishes to borrow by entering into a repo agreement to sell a bond today with a market value of 1 million dollars and repurchase it back **90 days later** (**the repurchase date**). The lender requires a repo rate of 2% and initial margin of 103%.

The initial purchase price (*the initial loan amount*) is calculated as the market value of the securities posted as collateral divided by the initial margin:

* $\qquad\text{Purchase Price (loan amount)} = \large\frac{\text{market value of securities}}{\text{initial margin}} = \frac{1,000,000}{1.03} = 970,874$


In [4]:
#### CALCULATE 90 day REPO LOAN, INTEREST, and RETURN + PRINCIPAL


# Date conversion 90 days
dateconv2 = (90/360)

## Calculate Purchase Price (loan amount) & Repurchase price
mkt_value = 1000000
purchase_price90 = mkt_value / 1.03
repurchase_price90 = purchase_price90 * (1 + (0.02 * dateconv2)) 

## Calculate Haircut
haircut90 = (mkt_value - purchase_price90) / mkt_value

## Calculate the interest on 90 days
int90 = repurchase_price90 - purchase_price90

#print("Market Value Price (loan amount) in dollars = $", round(mkt_value,0))
print("Purchase Price (loan amount) in dollars      = $", round(purchase_price90,0))
print("Repurchase Price (loan amount) in dollars    = $", round(repurchase_price90,0))
print("Interest paid on the 90 day loan in dollars  = $", round(int90,2))
print("Haircut percentage on the Purchase price     = %", round(haircut90,4))

Purchase Price (loan amount) in dollars      = $ 970874.0
Repurchase Price (loan amount) in dollars    = $ 975728.0
Interest paid on the 90 day loan in dollars  = $ 4854.37
Haircut percentage on the Purchase price     = % 0.0291


The repurchase price after 90 days is calculated as the purchase price (loan amount) multiplied by one plus the de-annualized repo rate. Assuming a 360-day count convention, this is calculated as follows for our example:

$\qquad\text{repurchase price} =  970,874 \times \big[1 + (0.02 \times (\frac{90}{360})\big] = 975,728$

The discount applied to the market value of collateral to get the purchase price is referred to as a haircut. In this example, the haircut is calculated as follows:

$\qquad\text{haircut} = \frac{(1,000,000 – 970,874)}{1,000,000} = 2.91\%$

The loan value increases during its life at the repo rate. Should the market value of the collateral fall below this value times the initial margin, the lender will ask the borrower for more collateral, known as variation margin.

For example, given the repo details just listed, assume that after 30 days the market value of the bond has fallen to 990,000 dollars:

adjusted loan amount after 

$\text{30 days} = 970,874 \times \Big[1 + (0.02 \times (\frac{30}{360})\Big] = 972,492$

                =  972,492 × 1.03 = 1,001,667
     
Because the collateral value is less than this, the borrower must provide **variation margin**.

    variation margin = (initial margin × adjusted purchase price) – market value of collateral

                = 1,001,667 - 990,000 = 11,667

In [5]:
#### Calculate the loan variation margin of 30 days below mkt value of collateral

## Set Date conversion
dateconv3 = 30/360

## Adjusted loan amount after 30 days
var30_loan_amt = purchase_price90 * (1 + (0.02 * dateconv3))
print("The Adjusted loan amount after 30 days in dollars     = $", round(var30_loan_amt,0))

## Because the collateral value is less than this, the borrower must provide variation margin
var30_margin = var30_loan_amt * 1.03
print("The Variation loan amount after 30 days in dollars    = $", round(var30_margin,0))

## Calculate the Variation Margin amount for a 30-day period when the collateral falls below $990,000
var30_below = var30_margin-990000 
print("The Variation margin of 30-days when falls below 990k = $", round(var30_below),0)

The Adjusted loan amount after 30 days in dollars     = $ 972492.0
The Variation loan amount after 30 days in dollars    = $ 1001667.0
The Variation margin of 30-days when falls below 990k = $ 11667 0
